File: scripts/vendor-rui-css.mjs
#!/usr/bin/env node
// Turn raft-ui's Tailwind token layer (dist/styles.css) into plain CSS for the
// static docs site. Tailwind-only at-rules are dropped; `@theme` blocks become
// `:root` so the palette variables the theme scopes reference still resolve.
//
//   npm pack raft-ui && tar xzf raft-ui-*.tgz
//   node scripts/vendor-rui-css.mjs package/dist/styles.css docs/assets/rui.css
import { readFile, writeFile } from 'node:fs/promises';
const [src, out] = process.argv.slice(2);
if (!src || !out) throw new Error('usage: vendor-rui-css.mjs <styles.css> <out.css>');
let css = await readFile(src, 'utf8');
const version = JSON.parse(await readFile(new URL('../package.json', `file://${process.cwd()}/${src}`), 'utf8').catch(() => '{}')).version ?? 'unknown';
function dropBlock(text, opener) {
  let i = text.indexOf(opener);
  while (i !== -1) {
    let depth = 0, j = text.indexOf('{', i);
    for (; j < text.length; j++) { if (text[j] === '{') depth++; else if (text[j] === '}' && --depth === 0) break; }
    text = text.slice(0, i) + text.slice(j + 1);
    i = text.indexOf(opener);
  }
  return text;
}
css = dropBlock(css, '@custom-variant');
css = dropBlock(css, '@utility');
css = css.replace(/@theme inline \{/g, ':root {').replace(/@theme \{/g, ':root {');
const leftover = css.replace(/\/\*[\s\S]*?\*\//g, "").match(/@(?!media|import|font-face|supports|keyframes)[a-z-]+/g);
if (leftover) throw new Error(`unhandled at-rules: ${[...new Set(leftover)].join(', ')}`);
await writeFile(out, `/* raft-ui ${version} token layer, flattened for static HTML by scripts/vendor-rui-css.mjs. Do not edit by hand. */\n${css}`);
console.log(`${out}: ${css.length} bytes`);

File: scripts/vendor-rui-css.py
#!/usr/bin/env python3
"""Flatten a downloaded raft-ui package's token stylesheet for static HTML.

Usage: python3 scripts/vendor-rui-css.py package/dist/styles.css docs/assets/rui.css
"""
import json
import re
import sys
from pathlib import Path

if len(sys.argv) != 3:
    raise SystemExit(__doc__)
source, output = map(Path, sys.argv[1:])
css = source.read_text(encoding="utf-8")
manifest = source.parent.parent / 'package.json'
version = json.loads(manifest.read_text(encoding="utf-8")).get('version', 'unknown') if manifest.exists() else 'unknown'
for opener in ('@custom-variant', '@utility'):
    while opener in css:
        start = css.index(opener)
        brace = css.find('{', start)
        if brace < 0:
            raise SystemExit('unterminated block: ' + opener)
        depth = 0
        for end in range(brace, len(css)):
            depth += (css[end] == '{') - (css[end] == '}')
            if depth == 0:
                css = css[:start] + css[end + 1:]
                break
        else:
            raise SystemExit('unterminated block: ' + opener)
css = re.sub(r'@theme(?: inline)? \{', ':root {', css)
leftovers = re.findall(r'@(?!media|import|font-face|supports|keyframes)[a-z-]+', re.sub(r'/\*.*?\*/', '', css, flags=re.S))
if leftovers:
    raise SystemExit('unhandled at-rules: ' + ', '.join(sorted(set(leftovers))))
output.write_text(f'/* raft-ui {version} token layer, flattened for static HTML by scripts/vendor-rui-css.py. Do not edit by hand. */\n' + css, encoding='utf-8')
print(f'{output}: {len(css)} characters')
